### PR TITLE
agent: add support on Node 12 and lower

### LINKED ISSUE #529: Agent without WeakRef dependency
It should be possible to implement `Agent` without using weak refs by using the `'connect'`/`'disconnect'` events.

### PR DESCRIPTION
**Pull request by dnlup**
Use `Pool`'s `connect`/`disconnect` events to add support of the `Agent` on Node 12 and lower.

Fixes #529 

Note:

I think it's better to wait for https://github.com/nodejs/undici/pull/521 to get merged before continuing. I have to touch many of the same blocks of that PR.

### CODE REVIEW

**Comment by Ethan-Arrowood** [comment]
@dnlup if you have the time, I could use a hand closing my PR. I'm struggling to get the tests to behave like I want them to and I can't seem to crack it

**Comment by dnlup** [comment]
> @dnlup if you have the time, I could use a hand closing my PR. I'm struggling to get the tests to behave like I want them to and I can't seem to crack it

Ok, I'll see what I can do

File: lib/agent.js [hunk 1]
```diff
@@ -72,6 +83,7 @@ class Agent {
   }
 }
 
+Agent.prototype.get = util.isWeakRefsSupported ? getFromWeakCache : getFromCache
```
Comment (ronag) [inline lib/agent.js:86]: I think we can skip the weak cache version if this works?

Comment (dnlup) [reply lib/agent.js:86]: Sounds good, I'll leave the logic for now but use only the version without the weakCache and see what happens.

Comment (dnlup) [reply lib/agent.js]: It looks like it works, so I think we can switch to a version without `WeakRefs` easily.

File: lib/agent.js [hunk 2]
```diff
@@ -72,13 +71,61 @@ class Agent {
...
+  constructor (opts) {
+    this[kAgentOpts] = opts
+    this[kAgentCache] = new Map()
+    if (util.isWeakRefsSupported) {
+      initWeakCache(this)
+    }
+  }
+}
+
+if (util.isWeakRefsSupported) {
```
[the remark this answers is not in the document]
Comment (dnlup) [reply lib/agent.js]: Ok, I removed the `WeakRef` logic. I am missing the reason for using the `connect` event, though. Once a `Pool` instance has no clients connected, it gets removed from the cache. When requesting a pool, a new one is initialized if it has not been created before for the given `origin`. What's the key point I am not getting here?

File: lib/agent.js [hunk 3]
```diff
@@ -1,83 +1,57 @@
...
+    let pool = self[kAgentCache].get(origin)
 
-    // if one exists in the cache try to return the WeakRef
-    if (ref !== undefined) {
-      const cached = ref.deref()
-      if (cached !== undefined) {
-        return cached
+    function onDisconnect () {
+      if (this.connected === 0) {
+        this.removeListener('disconnect', onDisconnect)
```
Comment (ronag) [inline lib/agent.js:23]:
```suggestion
        this.off('disconnect', onDisconnect)
```

**Comment by ronag** [comment]
Don't forget to update docs

File: lib/agent.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcollina) [reply lib/agent.js]: Nothing, using `disconnect` is just fine.

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/agent.js [hunk 4]
```diff
@@ -1,83 +1,57 @@
...
+    const self = this
+    let pool = self[kAgentCache].get(origin)
 
-    // if one exists in the cache try to return the WeakRef
-    if (ref !== undefined) {
-      const cached = ref.deref()
-      if (cached !== undefined) {
-        return cached
+    function onDisconnect () {
+      if (this.connected === 0) {
```
Comment (ronag) [inline lib/agent.js:22]:
```suggestion
      if (this.connected === 0 && this.pending === 0) {
```

Comment (dnlup) [reply lib/agent.js:22]: Thanks, great catch

**Comment by ronag** [comment]
Done?

**Comment by dnlup** [comment]
> Done?

Yup, unless I missed something else

**Comment by ronag** [comment]
@dnlup This commit seems to cause CI failures. Could you investigate?

**Comment by dnlup** [comment]
Sure, are you referring to the failures https://github.com/nodejs/undici/pull/535 ?

**Comment by ronag** [comment]
Screen Shot 2021-02-05 at 14 21 19

**Comment by dnlup** [comment]
Ok, thanks

**Comment by dnlup** [comment]
That specific error was caused by a failure during the setup phase (it couldn't download the `checkout` action). I have seen similar errors [here](https://github.com/nodejs/undici/runs/1838763598?check_suite_focus=true). I'll dig some more to find failures related to this commit.

**Comment by ronag** [comment]
Ah I see. I'll just try re-running it then.

**Pull request merged**